File: 0036-valid-sudoku/solution.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int,set<char>> cols;
        unordered_map<int,set<char>> rows;
        unordered_map<int,set<char>> sub;

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                int val = board[r][c];
                if (val == '.') continue;

                int x = (r/3)*3 + (c/3);
                if ((rows[r].find(val) != rows[r].end()) || (cols[c].find(val) != cols[c].end()) || (sub[x].find(val) != sub[x].end())) {
                    return false;
                }

                // update hashmap
                rows[r].insert(val);
                cols[c].insert(val);
                sub[x].insert(val);
            }
        }

        return true;
    }
};

File: 0036-valid-sudoku/solution.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = defaultdict(set) # 0 to N.
        rows = defaultdict(set)
        boxs = defaultdict(set) # (x // 3)

        # r 0 -> 2 (box = 0,1,2)
        # r 3 -> 5 (box = 3,4,5)
        # r 6 -> 8 (box = 6,7,8)

        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue
                
                if board[r][c] in rows[r] or board[r][c] in cols[c] or board[r][c] in boxs[(r // 3, c // 3)]:
                    return False
                
                rows[r].add(board[r][c])
                cols[c].add(board[r][c])
                boxs[(r // 3, c // 3)].add(board[r][c])
        
        return True
